Lib/ctypes/util.py

```diff
@@ -306,10 +306,42 @@ def _findLib_ld(name):
                 pass  # result will be None
             return result
 
+        def _findLib_walkldpath(name):
+            # See issue #21622
+            def _is_elf(filepath):
+                try:
+                    with open(filepath, 'rb') as fh:
+                        return fh.read(4) == b'\x7fELF'
+                except:
+                    return False
+
+            from glob import glob
+            if os.path.isabs(name):
+                return name
+            if name.startswith("lib") and name.endswith(".so"):
+                suffixes = [ '.[0-9]*' ]
+                libname = name
+            else:
+                suffixes = ['.so', '.so.[0-9]*', '.musl-*.so.[0-9]*']
+                libname = 'lib'+name
```

Comment:
I believe the use of the `lib` prefix here is incorrect. Shared libraries are not required to be prefixed by `lib`. One example of a project that does not do this would be OpenMPI, which installs `mpi.so`. `find_library('mpi')` would try to find `libmpi.so` and fail. I think you should always search for `{name}.so` first and only after try to get guess the soname.

Reply:
Where can i find this `mpi.so`? In [alpine]( https://pkgs.alpinelinux.org/contents?branch=edge&name=openmpi&arch=x86_64&repo=testing) its `/usr/lib/libmpi.so.40` and in [debian](https://packages.debian.org/buster/amd64/libopenmpi3/filelist) it is `/usr/lib/x86_64-linux-gnu/libmpi.so.40`

Reply:
Sorry, I meant Boost MPI, not OpenMPI :sweat_smile:

Reply:
Do you have an exact URL or exact debian, ubuntu or fedora package? I still cannot find it. From what I can see [debian](https://packages.debian.org/sid/amd64/libboost-mpi1.71.0/filelist) uses `/usr/lib/x86_64-linux-gnu/libboost_mpi.so.1.71.0`.

I doubt that there exist any `{name}.so` that current `find_library` picks up, but if I'm wrong, I'd like to make a simple test case that works on a glibc system and make it work with musl as well.